File: dynuiparts/src/main/java/com/dylan/common/rx/RxBus2.java
package com.dylan.common.rx;


import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import io.reactivex.internal.operators.flowable.FlowableOnBackpressureError;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.processors.PublishProcessor;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;

public class RxBus2 {

    private static RxBus2 instance;
    private Subject<Object> subjectBus;
    private FlowableProcessor<Object> processorBus;

    public static RxBus2 getDefault() {
        if (instance == null) {
            synchronized (RxBus2.class) {
                if (instance == null) {
                    RxBus2 tempInstance = new RxBus2();
                    tempInstance.subjectBus = PublishSubject.create().toSerialized();
                    tempInstance.processorBus = PublishProcessor.create().toSerialized();
                    instance = tempInstance;
                }
            }
        }
        return instance;
    }

    public <T> CompositeDisposable register(Class<T> eventType, Consumer<T> observer, CompositeDisposable disposable) {
        if (disposable == null) disposable = new CompositeDisposable();
        disposable.add(toObserverable(eventType).observeOn(AndroidSchedulers.mainThread()).subscribe(observer));
        return disposable;
    }

    public <T> Disposable register(Class<T> eventType, Consumer<T> observer) {
        return toObserverable(eventType).observeOn(AndroidSchedulers.mainThread()).subscribe(observer);
    }
    public <T> Disposable register(Class<T> eventType, Consumer<T> observer, Scheduler scheduler) {
        return toObserverable(eventType).observeOn(scheduler).subscribe(observer);
    }
    public <T> Disposable register(Class<T> eventType, Consumer<T> observer, Scheduler scheduler, BackpressureStrategy strategy) {
        Flowable o = toFlowable(eventType);
        switch (strategy) {
            case DROP:
                o = o.onBackpressureDrop();
            case LATEST:
                o = o.onBackpressureLatest();
            case MISSING:
                o = o;
            case ERROR:
                o = RxJavaPlugins.onAssembly(new FlowableOnBackpressureError<>(o));
            default:
                o = o.onBackpressureBuffer();
        }
        if (scheduler != null) {
            o.observeOn(scheduler);
        }
        return o.subscribe(observer);
    }

    public <T> Disposable register(Class<T> eventType, Consumer<T> observer, BackpressureStrategy strategy) {
        return register(eventType, observer, null, strategy);
    }

    public void unregister(Disposable disposable) {
        if (disposable != null && !disposable.isDisposed()) {
            disposable.dispose();
        }
    }

    public void unregister(CompositeDisposable compositeDisposable) {
        if (compositeDisposable != null) {
            compositeDisposable.dispose();
        }
    }

    public void post(final Object event) {
        subjectBus.onNext(event);
        processorBus.onNext(event);
    }

    private Observable toObserverable(Class cls) {
        return subjectBus.ofType(cls);
    }

    private Flowable toFlowable(Class cls) {
        return processorBus.ofType(cls);
    }

    public boolean hasObservers() {
        return subjectBus.hasObservers();
    }

    public boolean hasSubscribers() {
        return processorBus.hasSubscribers();
    }
}

File: dynkotlins/src/main/java/com/yunsean/dynkotlins/rx/RxBus2.kt
package com.yunsean.dynkotlins.rx

import io.reactivex.Observable
import io.reactivex.ObservableTransformer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer
import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.Subject


class RxBus2 {
    private var subjectBus: Subject<Any> = PublishSubject.create<Any>().toSerialized()

    fun <T, R> register(eventType: Class<T>, composer: ObservableTransformer<in T?, out R>?, observer: Consumer<R>): RxBus2 {
        toObserverable(eventType)
            .compose(composer)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(observer)
        return this
    }

    fun <T> register(eventType: Class<T>, observer: Consumer<T>, disposable: CompositeDisposable?): CompositeDisposable {
        var disposable = disposable
        if (disposable == null) disposable = CompositeDisposable()
        disposable.add(toObserverable<T>(eventType).observeOn(AndroidSchedulers.mainThread()).subscribe(observer))
        return disposable
    }
    fun <T> register(eventType: Class<T>, observer: Consumer<T>): Disposable {
        return toObserverable<T>(eventType).observeOn(AndroidSchedulers.mainThread()).subscribe(observer)
    }
    fun unregister(disposable: Disposable?) {
        if (disposable != null && !disposable.isDisposed) {
            disposable.dispose()
        }
    }
    fun unregister(compositeDisposable: CompositeDisposable?) {
        compositeDisposable?.clear()
    }
    fun post(event: Any) {
        subjectBus.onNext(event)
    }
    private fun <T> toObserverable(cls: Class<T>): Observable<T> = subjectBus.ofType(cls)
    fun hasObservers(): Boolean = subjectBus.hasObservers()

    private object Holder {
        val INSTANCE = RxBus2()
    }
    companion object {
        val shared: RxBus2 by lazy { Holder.INSTANCE }
    }
}
